python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -636,6 +637,201 @@ def assets_defs_by_key(self) -> Mapping[AssetKey, AssetsDefinition]:
         return self._assets_by_key
 
 
+@experimental
+class AssetSLASensorEvaluationContext(SensorEvaluationContext):
+    """The context object available as the argument to the evaluation function of a :py:class:`dagster.AssetSLASensorDefinition`."""
+
+    def __init__(
+        self,
+        instance_ref: Optional[InstanceRef],
+        last_completion_time: Optional[float],
+        last_run_key: Optional[str],
+        cursor: Optional[str],
+        repository_name: Optional[str],
+        repository_def: "RepositoryDefinition",
+        asset_keys: Optional[Sequence[AssetKey]] = None,
+        instance: Optional[DagsterInstance] = None,
+    ):
+        from dagster._core.selector.subset_selector import generate_asset_dep_graph
+
+        self._repository_def = repository_def
+
+        self._upstream_assets = generate_asset_dep_graph(
+            self._repository_def._assets_defs_by_key.values(),
+            self._repository_def.source_assets_by_key.values(),
+        )["upstream"]
+
+        self._assets_by_key: Dict[AssetKey, AssetsDefinition] = {}
+        # if no asset_keys supplied, default to all
+        for asset_key in asset_keys or repository_def._assets_defs_by_key.keys():
+            assets_def = self._repository_def._assets_defs_by_key.get(asset_key)
+            if assets_def is None:
+                raise DagsterInvalidDefinitionError(
+                    f"No asset with {asset_key} found in repository"
+                )
+            if not assets_def.slas_by_key.get(asset_key):
+                continue
+            self._assets_by_key[asset_key] = assets_def
+
+        self._next_cursor = cursor
+
+        super(AssetSLASensorEvaluationContext, self).__init__(
+            instance_ref=instance_ref,
+            last_completion_time=last_completion_time,
+            last_run_key=last_run_key,
+            cursor=cursor,
+            repository_name=repository_name,
+            instance=instance,
+        )
+
+    def _data_times_by_key(self) -> Mapping[AssetKey, Optional[float]]:
+        """
+        For each asset monitored by this sensor, calculate its "data time". The data time of an
+        asset is the time of the earliest data that could be incorporated into the current state
+        of the asset, calculated for each asset by comparing their latest timestamp to the latest
+        parent timestamp that happened before that materialization (as you cannot incorporate parent
+        data if that parent runs after you do).
+        """
+        from dagster._core.events import DagsterEventType
+        from dagster._core.storage.event_log.base import EventRecordsFilter
+
+        calculated_data_times: Dict[Tuple[AssetKey, Optional[float]], Optional[float]] = {}
```

Comment:
we're using a little bit of dynamic programming here -- this isn't just for show, as in a lot of normal scenarios there's a lot of overlap between query paths that need to be taken up the tree. might do some benchmarking on this to see if it helps in practice